src/main/java/com/fauna/client/FaunaClient.java

```diff
@@ -42,10 +46,38 @@ protected String getFaunaSecret() {
         return this.faunaSecret;
     }
 
+    public Optional<Long> getLastTransactionTs() {
+        long ts = lastTransactionTs.get();
+        return ts > 0 ? Optional.of(ts) : Optional.empty();
+    }
+
     private static <T> Supplier<CompletableFuture<QuerySuccess<T>>> makeAsyncRequest(HttpClient client, HttpRequest request, Codec<T> codec) {
         return () -> client.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream()).thenApply(body -> QueryResponse.parseResponse(body, codec));
     }
 
+    private static Optional<ServiceException> extractServiceException(Throwable throwable) {
+        if (throwable.getCause() instanceof FaunaException) {
+            return Optional.of((ServiceException) throwable.getCause());
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private void updateTs(QueryResponse resp) {
+        Long newTs = resp.getLastSeenTxn();
+        if (newTs != null) {
+            this.lastTransactionTs.updateAndGet(oldTs -> newTs > oldTs ? newTs : oldTs );
+        }
+    }
+
+    private <T> void completeRequest(QuerySuccess<T> success, Throwable throwable) {
+        if (success != null) {
+            updateTs(success);
```

Comment:
I think we need to update the last seen txn for both success and failures according to our open API spec.